### PR TITLE
8309475: Test java/foreign/TestByteBuffer.java fails: a problem with msync (aix)

### PR DESCRIPTION
**Pull request by backwaterred**
This change adds additional support to MappedByteBuffer::force which delegates to msync on AIX. Specifically, it checks whether and error with errno set to EINVAL is the cause of a msync call to an address mmapped with the MAP_PRIVATE flag set. If this flag is set, then the msync call EINVAL is expected, and can be ignored as per the Java documentation of the force method.

> In particular, the method has no effect for buffers mapped in read-only or private mapping modes. [1]

Separated into separate file as per recommendation by @AlanBateman.

[1] https://download.java.net/java/early_access/jdk21/docs/api/java.base/java/nio/MappedByteBuffer.html#force()

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309475](https://bugs.openjdk.org/browse/JDK-8309475): Test java/foreign/TestByteBuffer.java fails: a problem with msync (aix) (**Bug** - P4)

### Reviewers
 * @TOatGithub (no known openjdk.org user name / role) ⚠️ Review applies to [35657914](https://git.openjdk.org/jdk/pull/14964/files/3565791466f547c33f918e29fff9754e12cce4e5)
 * @JoKern65 (no known openjdk.org user name / role) ⚠️ Review applies to [35657914](https://git.openjdk.org/jdk/pull/14964/files/3565791466f547c33f918e29fff9754e12cce4e5)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [9bb09167](https://git.openjdk.org/jdk/pull/14964/files/9bb09167492833de6b893bb23a29e297f7579599)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [9bb09167](https://git.openjdk.org/jdk/pull/14964/files/9bb09167492833de6b893bb23a29e297f7579599)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14964/head:pull/14964` \
`$ git checkout pull/14964`

Update a local copy of the PR: \
`$ git checkout pull/14964` \
`$ git pull https://git.openjdk.org/jdk.git pull/14964/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14964`

View PR using the GUI difftool: \
`$ git pr show -t 14964`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14964.diff">https://git.openjdk.org/jdk/pull/14964.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14964#issuecomment-1649957221)

### CODE REVIEW

**Comment by AlanBateman** [comment]
It's good that this doesn't clutter the Unix implementation of these methods with a lot of AIX-specific code. However, part of me wonders if force should special case the read-only or private mapping cases so that it's just a no-op, need to think about that one as the MBB implementation doesn't know if the mapping is read-write or private.

**Comment by backwaterred** [comment]
> part of me wonders if force should special case the read-only or private mapping cases so that it's just a no-op, need to think about that one as the MBB implementation doesn't know if the mapping is read-write or private.

I agree with the sentiment. It would be nice to have this information available without so much heavy lifting.

**Comment by RealCLanger** [comment]
FYI: We are excluding the failing test for the time being in #15055. The exclusion will need to be removed with the final commit for this.

**Comment by backwaterred** [comment]
Thanks for mentioning it @RealCLanger. I'll remove the exception from the ProblemList before I merge.

This PR and #14963 are ready for reviews if anyone has a chance to take a look.

**Comment by backwaterred** [comment]
Fixing chaos brought on by merging with upstream/master branch.

**Comment by TOatGithub** [comment]
Code seems reasonable: on msync error EINVAL retrieves info about flags of mapped region and reacts accordingly.

**Comment by backwaterred** [comment]
Thanks Thomas :smile:

**Comment by JoKern65** [comment]
I can only agree with Thomas that this is a reasonable and also inspiring solution to the problem.
Very good job Tyler.
Please reviewers take over and do your job (@MBaesken, @TheRealDoerr)

**Comment by MBaesken** [comment]
Hi Tyler, I added the PR to our internal  build/test queue. Will check the results later.

**Review by TOatGithub — APPROVED** [review]
Code seems reasonable: on msync error EINVAL retrieves info about flags of mapped region and reacts accordingly.

**Review by JoKern65 — APPROVED** [review]
This is a reasonable solution for the problem.

File: src/java.base/aix/native/libnio/MappedMemoryUtils.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+{
+    while (!feof(proc_file)) {
+        memset(map_entry, '\0', sizeof(prmap_t));
+        fread((char*)map_entry, sizeof(prmap_t), 1, proc_file);
+        if (ferror(proc_file)) {
+            JNU_ThrowIOExceptionWithMessageAndLastError(env,
+                        "msync with parameter MS_SYNC failed (could not read /proc/<pid>/map)");
+            return;
+        } else if (map_entry->pr_vaddr <= end_address &&
+                   end_address <= map_entry->pr_vaddr + map_entry->pr_size) {
```
Comment (TheRealMDoerr) [inline src/java.base/aix/native/libnio/MappedMemoryUtils.c:140]:
```
MappedMemoryUtils.c:140:55: error: arithmetic on a pointer to void is a GNU extension [-Werror,-Wgnu-pointer-arith]
                   end_address <= map_entry->pr_vaddr + map_entry->pr_size) {
                                  ~~~~~~~~~~~~~~~~~~~ ^
```

**Comment by MBaesken** [comment]
The compile error above mentioned by Martin is observed with xlC 17.1.1   (build with xlc16 still works).

**Review by MBaesken — APPROVED** [review]
Otherwise looks okay but please fix the xlc17 build before pushing.

File: src/java.base/aix/native/libnio/MappedMemoryUtils.c (hunk 1 above)
Comment (backwaterred) [reply src/java.base/aix/native/libnio/MappedMemoryUtils.c:140]: Thanks for finding this. I am still building and testing with xlc 16.1, but it's a good idea to add 17.1 to the standard process as well. I'll fix before merging.

**Comment by backwaterred** [comment]
There are some issues with the integration tests apparently. Those failures do not appear related to my changes:
```
Package g++-10 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
```

File: src/java.base/aix/native/libnio/MappedMemoryUtils.c (hunk 1 above)
Comment (TheRealMDoerr) [reply src/java.base/aix/native/libnio/MappedMemoryUtils.c]: Thanks for fixing!

Comment (TheRealMDoerr) [reply src/java.base/aix/native/libnio/MappedMemoryUtils.c]:
```
MappedMemoryUtils.c:140:32: error: ordered comparison between pointer and integer ('void *' and 'unsigned long') [-Werror]
                   end_address <= (uint64_t)map_entry->pr_vaddr + map_entry->pr_size) {
                   ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by TheRealMDoerr** [comment]
`(uint64_t)end_address <= (uint64_t)map_entry->pr_vaddr + map_entry->pr_size` works with xlc 17.

**Comment by backwaterred** [comment]
> (uint64_t)end_address <= (uint64_t)map_entry->pr_vaddr + map_entry->pr_size works with xlc 17.

I thought it might complain about that as well. I made this change.

**Review by AlanBateman — APPROVED** [review]
The removal of the AIX specific code from the Unix version of MappedMemoryUtils.c look fine. At some point I think we should re-visit this as the force method should be a no-op for the read-only or private mapping cases so this code should not be needed.

**Comment by MBaesken** [comment]
The build now works as well with xlc17.

**Review by MBaesken — APPROVED** [review]

**Comment by TheRealMDoerr** [comment]
I think you need to remove "java/foreign/TestByteBuffer.java 8309475 aix-ppc64" from test/jdk/ProblemList.txt.

**Comment by backwaterred** [comment]
Good point. I did that after the last merge too :-)

**Pull request closed without merging**